### PR TITLE
initial traced implementation

### PR DESCRIPTION
**Pull request by vyzo**
Depends on https://github.com/libp2p/go-libp2p-pubsub/pull/227

### CODE REVIEW

**Comment by vyzo** [comment]
summoning @Stebalien @raulk 
This is ready to be used and itching to get into lotus!

**Comment by vyzo** [comment]
Sample run from `tracestat` on a trace generated by running a go test with 20 nodes, publishing 1k messages:

```
$ tracestat -cdf /tmp/traced.out/trace.15741*
=== Trace Summary ===
Peers: 20
Published Messages: 1000
Delivered Messages: 20000
Duplicate Messages: 111292
Rejected Messages: 0
Sent RPCs: 139976
Dropped RPCS: 0
=== Propagation Delay CDF (ms) ===
0 90
1 1123
2 3763
3 5859
4 8328
5 10772
6 13055
7 15140
8 16813
9 18114
10 19021
11 19506
12 19708
13 19785
14 19839
15 19867
16 19885
17 19894
18 19907
19 19915
20 19918
21 19925
22 19927
23 19931
24 19938
25 19941
26 19942
27 19944
28 19946
29 19947
30 19951
31 19952
32 19954
34 19955
36 19956
37 19961
38 19967
39 19972
40 19977
41 19982
42 19986
43 19992
44 19996
45 19998
46 19999
47 20000
```

File: cmd/tracestat/main.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+			log.Printf("error decoding trace event from %s: %s", f, err)
+			return
+		}
+
+		ts.addEvent(&evt)
+	}
+}
+
+func (ts *tracestat) addEvent(evt *pb.TraceEvent) {
+	peer := peer.ID(evt.GetPeerID())
```
[the remark this answers is not in the document]
Comment (vyzo) [reply cmd/tracestat/main.go]: No, it's raw bytes and can be coerced directly.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,54 @@
...
+
+```
+$ git clone git@github.com:libp2p/go-libp2p-pubsub-tracer.git
+$ cd go-libp2p-pubsub-tracer
+$ go install ./...
+```
+
+## Tools
+
+### traced
```
[the remark this answers is not in the document]
Comment (vyzo) [reply README.md]: I disagree. `collected` is so generic as to be meaningless. What does it collect? Why, traces, so it's a trace daemon, aka `traced`.

File: tracer.go [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+	for {
+		select {
+		case <-t.wr:
+		case <-t.flush:
+			flush = true
+		}
+
+		t.mx.Lock()
+		tmp := t.buf
```
[the remark this answers is not in the document]
Comment (vyzo) [reply tracer.go]: it's not a bug, that's exactly what we want.

[the remark this answers is not in the document]
Comment (vyzo) [reply tracer.go]: Note that the buffers are swapped.

[the remark this answers is not in the document]
Comment (raulk) [reply tracer.go]: What the assignment does is copy the [slice descriptor](https://blog.golang.org/go-slices-usage-and-internals#TOC_4.).

Then this code resizes the slice to be of length 0, but the start address and the capacity remain intact.

So after you release the lock, any new items that arrive will be written in the original memory segments, and those writes will be visible in tmp/buf.

[the remark this answers is not in the document]
Comment (vyzo) [reply tracer.go]: it's two different slices (buffers). We swapped between one and the other, to keep the allocated slice and avoid the copy.
It's just double buffering.

[the remark this answers is not in the document]
Comment (raulk) [reply tracer.go]: Ok, so this thread is moot. There are two different slices here: the struct one (`t.buf`) and a local one (`buf`). This code swaps the slice descriptors, so there's no risk of dirty writes.

**Comment by vyzo** [comment]
This is about the expected amplification with the default parameters.
Regarding the visualization, there is work ongoing to build a visualizer afaik.

**Pull request merged**